### PR TITLE
Access control: Extend GetUserPermissions() to query permissions in org

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What is this feature?**

Extend GetUserPermissions() method to be able to get permissions in a specific org.

**Why do we need this feature?**

We need this to simplify code and avoid copying of signed user object.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/api/org_users.go [hunk 1]
```diff
@@ -323,12 +323,13 @@ func (hs *HTTPServer) searchOrgUsersHelper(c *contextmodel.ReqContext, query *or
 
 	// Get accesscontrol metadata and IPD labels for users in the target org
 	accessControlMetadata := map[string]accesscontrol.Metadata{}
-	if c.QueryBool("accesscontrol") && c.SignedInUser.Permissions != nil {
-		// TODO https://github.com/grafana/identity-access-team/issues/268 - user access control service for fetching permissions from another organization
-		permissions, ok := c.SignedInUser.Permissions[query.OrgID]
-		if ok {
-			accessControlMetadata = accesscontrol.GetResourcesMetadata(c.Req.Context(), permissions, "users:id:", userIDs)
+	if c.QueryBool("accesscontrol") {
+		permissionsList, err := hs.accesscontrolService.GetUserPermissions(c.Req.Context(), c.SignedInUser, accesscontrol.Options{OrgID: &query.OrgID})
```
Comment (Jguer) [inline pkg/api/org_users.go:327]: is it ok to not rely on any caching here?

Not blocking

**Review by Jguer — APPROVED** [review]

**Comment by gamab** [comment]
Haven't reviewed this, but are their chances of getting users permissions in orgs you should not be allowed to?

File: pkg/api/org_users.go (hunk 1 above)
Comment (alexanderzobnin) [reply pkg/api/org_users.go:327]: What do you mean? `GetUserPermissions()` uses cache if it's enabled. Do you suggest to disable it here?

**Comment by alexanderzobnin** [comment]
@gamab not allowed to list permissions or what? I think user already has everything under `SignedInUser.Permissions` struct, so this is just another way to get it. But maybe I miss something.

**Comment by gamab** [comment]
Nevermind, I read the PR and thought it was something different.

**Review by gamab — APPROVED** [review]

**Review by gamab — CHANGES_REQUESTED** [review]
On second thought, we cannot just optionally change the `orgID`, we need to change the role and user's teams as well.

**Comment by alexanderzobnin** [comment]
Yes, that's true, good catch! I'll dig into that.

**Comment by alexanderzobnin** [comment]
@gamab I'm thinking about using similar SQL query to get information about user permissions in specific org. The main issue with `identity.Requester` approach used in GetUserPermissions is that it uses basic role and list of teams that assigned on login. This information is not valid for different org, so we need to fetch it again. So the DB query is the most straightforward way to do it and we can make sure we correctly fetch basic role and team-associated roles. This query is 80% the same as you use for plugins permissions, so we can reuse it (you mentioned that role filter might be useful for plugins as well).

File: pkg/services/accesscontrol/database/database.go [hunk 2]
```diff
@@ -53,6 +54,65 @@ func (s *AccessControlStore) GetUserPermissions(ctx context.Context, query acces
...
+
+		params := []any{accesscontrol.RoleGrafanaAdmin, accesscontrol.GlobalOrgID, query.OrgID, query.UserID}
+
+		if len(query.RolePrefixes) > 0 {
+			q += " AND ( " + strings.Repeat("r.name LIKE ? OR ", len(query.RolePrefixes)-1)
+			q += "r.name LIKE ? )"
+			for _, prefix := range query.RolePrefixes {
+				params = append(params, prefix+"%")
+			}
+		}
```
Comment (gamab) [inline pkg/services/accesscontrol/database/database.go:106]: Any reason not to use SearchUsersPermissions with this additional filter?

Comment (alexanderzobnin) [reply pkg/services/accesscontrol/database/database.go:106]: No reason, just need to refactor it to be able to use filter.

Comment (gamab) [inline pkg/services/accesscontrol/database/database.go:106]: Any reason why we just don't add that to the `SearchUsersPermissions` function instead?

Comment (gamab) [reply pkg/services/accesscontrol/database/database.go:106]: I would refactor `SearchUsersPermissions` then to de-duplicate code, wdyt?

File: pkg/services/accesscontrol/database/database.go [hunk 3]
```diff
@@ -92,7 +92,8 @@ func (s *AccessControlStore) SearchUsersPermissions(ctx context.Context, orgID i
 					) AS sa ON 1 = 1
 					WHERE br.role = ?
 		) AS up
-		WHERE (org_id = ? OR org_id = ?)
+		INNER JOIN role AS r on up.role_id = r.id
```
Comment (gamab) [inline pkg/services/accesscontrol/database/database.go:95]: Oh I understand now why you preferred the other approach! It adds an additional join :confounded:

Comment (alexanderzobnin) [reply pkg/services/accesscontrol/database/database.go:95]: Yes, one little detail :) We actually need it only when filter by role name, so I'll try do add it under condition.

File: pkg/services/accesscontrol/database/database.go [hunk 4]
```diff
@@ -36,8 +36,8 @@ func (s *AccessControlStore) GetUserPermissions(ctx context.Context, query acces
 		` + filter
 
 		if len(query.RolePrefixes) > 0 {
-			q += " WHERE ( " + strings.Repeat("role.name LIKE ? OR ", len(query.RolePrefixes))
-			q = q[:len(q)-4] + " )" // remove last " OR "
```
Comment (alexanderzobnin) [inline pkg/services/accesscontrol/database/database.go:40]: Just want to fix this as well

File: pkg/services/accesscontrol/acimpl/service.go [hunk 5]
```diff
@@ -237,6 +264,8 @@ func (s *Service) DeclarePluginRoles(ctx context.Context, ID, name string, regs
 // SearchUsersPermissions returns all users' permissions filtered by action prefixes
 func (s *Service) SearchUsersPermissions(ctx context.Context, usr identity.Requester,
 	options accesscontrol.SearchOptions) (map[int64][]accesscontrol.Permission, error) {
+	// Limit roles to available in OSS
+	options.RolePrefixes = []string{accesscontrol.BasicRolePrefix, accesscontrol.ManagedRolePrefix, accesscontrol.ExternalServiceRolePrefix}
```
Comment (alexanderzobnin) [inline pkg/services/accesscontrol/acimpl/service.go:268]: Enterprise reuses the same `SearchUsersPermissions()` store method, so we cannot put filters there. So instead I added it to the service method, to make sure we only return permissions available in OSS. Enterprise method does not contain filters.

Comment (gamab) [inline pkg/services/accesscontrol/acimpl/service.go:268]:
```suggestion
	options.RolePrefixes = []string{accesscontrol.ManagedRolePrefix, accesscontrol.ExternalServiceRolePrefix}
```

File: pkg/services/accesscontrol/acimpl/service.go [hunk 6]
```diff
@@ -158,6 +159,32 @@ func (s *Service) getCachedUserPermissions(ctx context.Context, user identity.Re
 	return permissions, nil
 }
 
+func (s *Service) GetUserPermissionsInOrg(ctx context.Context, user identity.Requester, orgID int64) ([]accesscontrol.Permission, error) {
```
Comment (gamab) [inline pkg/services/accesscontrol/acimpl/service.go:162]: maybe add some tests?

**Review by gamab — APPROVED** [review]
LGTM!

File: pkg/services/accesscontrol/acimpl/service.go [hunk 7]
```diff
@@ -158,6 +159,32 @@ func (s *Service) getCachedUserPermissions(ctx context.Context, user identity.Re
...
+	namespace, id := user.GetNamespacedID()
+	userID, err := identity.UserIdentifier(namespace, id)
+	if err != nil {
+		return nil, err
+	}
+
+	dbPermissions, err := s.store.SearchUsersPermissions(ctx, orgID, accesscontrol.SearchOptions{
+		NamespacedID: authn.NamespacedID(namespace, userID),
+		// Query only basic, managed and plugin roles in OSS
+		RolePrefixes: []string{accesscontrol.BasicRolePrefix, accesscontrol.ManagedRolePrefix, accesscontrol.ExternalServiceRolePrefix},
```
[the remark this answers is not in the document]
Comment (alexanderzobnin) [reply pkg/services/accesscontrol/acimpl/service.go:178]: Ah, that's true... And that's a problem since we need to add it to the list of roles.

**Pull request merged**